### PR TITLE
[AutoDiff] add cross-file registration flag to .swiftinterface

### PR DESCRIPTION
**Pull request by marcrasi**
Fixes TF-1101 and https://github.com/tensorflow/swift/issues/359.

### CODE REVIEW

**Review by dan-zheng — APPROVED** [review]
Sweet! Could you please build a macOS toolchain to verify that the issue is fixed?

Edit: I'm fine with testing macOS toolchain builds after this PR is merged, too. If macOS toolchains are indeed fixed, we can directly release the built toolchain to [tensorflow/swift](https://github.com/tensorflow/swift/blob/master/Installation.md).

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by compnerd** [comment]
@dan-zheng https://dev.azure.com/compnerd/swift-build/_build/results?buildId=24025&view=results should do a build of the macOS toolchain

File: lib/Frontend/CompilerInvocation.cpp
Comment (compnerd) [inline lib/Frontend/CompilerInvocation.cpp]: Why is this `-Xllvm`?  Should this not be `-Xswift`?  This is part of the Swift side of things

Comment (dan-zheng) [reply lib/Frontend/CompilerInvocation.cpp]: Basically all SILOptimizer flags are declared via `llvm::cl::opt<T>`, which adds a new `-Xllvm`-prefixed flag.

I actually haven't heard of `-Xswift`. Is there a way to register command line options prefixed with `-Xswift` instead of `-Xllvm`?

**Comment by marcrasi** [comment]
> @dan-zheng https://dev.azure.com/compnerd/swift-build/_build/results?buildId=24025&view=results should do a build of the macOS toolchain

Is there a way that I can trigger that on this branch?

I already have one in progress on our mac desktop right now but the builds there seem quite flaky so it would be nice to trigger one there too in case the desktop fails. And it looks easier to build things there than in the desktop in the future.

**Comment by marcrasi** [comment]
> I already have one in progress on our mac desktop right now but the builds there seem quite flaky

Indeed, the build on the desktop just failed with some error message that looks unrelated to this PR:

```
--- Building swiftpm ---
+ /Users/swiftninjas/s4tf/swiftpm/Utilities/new-bootstrap build --release --swiftc-path /Users/swiftninjas/s4tf/swift/swift-nightly-install/Library/Developer/Toolchains/swift-tensorflow-DEVELOPMENT-2020-01-10-a.xctoolchain/usr/bin/swiftc --clang-path /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang --cmake-path /usr/local/bin/cmake --ninja-path /Users/swiftninjas/s4tf/build/buildbot_osx/ninja-build/ninja --build-dir /Users/swiftninjas/s4tf/build/buildbot_osx/swiftpm-macosx-x86_64 --llbuild-build-dir /Users/swiftninjas/s4tf/build/buildbot_osx/llbuild-macosx-x86_64
--- bootstrap: note: Building SwiftPM (with CMake)
[0/1][  0%][0.000s] Re-running CMake...
CMake Error at CMakeLists.txt:13 (project):
  The CMAKE_Swift_COMPILER:

    /Users/swiftninjas/s4tf/swift/swift-nightly-install/Library/Developer/Toolchains/swift-tensorflow-DEVELOPMENT-2019-12-22-a.xctoolchain/usr/bin/swiftc

  is not a full path to an existing compiler tool.

  Tell CMake where to find the compiler by setting either the environment
  variable "SWIFTC" or the CMake cache entry CMAKE_Swift_COMPILER to the full
  path to the compiler, or to the compiler name if it is in the PATH.


-- Configuring incomplete, errors occurred!
See also "/Users/swiftninjas/s4tf/build/buildbot_osx/swiftpm-macosx-x86_64/x86_64-apple-macosx/bootstrap/CMakeFiles/CMakeOutput.log".
See also "/Users/swiftninjas/s4tf/build/buildbot_osx/swiftpm-macosx-x86_64/x86_64-apple-macosx/bootstrap/CMakeFiles/CMakeError.log".
FAILED: build.ninja
/usr/local/Cellar/cmake/3.15.4/bin/cmake -S/Users/swiftninjas/s4tf/swiftpm -B/Users/swiftninjas/s4tf/build/buildbot_osx/swiftpm-macosx-x86_64/x86_64-apple-macosx/bootstrap
ninja: error: rebuilding 'build.ninja': subcommand failed
/Users/swiftninjas/s4tf/build/buildbot_osx/ninja-build/ninja
--- bootstrap: error: Command '['/Users/swiftninjas/s4tf/build/buildbot_osx/ninja-build/ninja']' returned non-zero exit status 1
./utils/build-script: fatal error: command terminated with a non-zero exit status 1, aborting
./utils/build-script: fatal error: command terminated with a non-zero exit status 1, aborting
```

I'll try building it on my mac laptop which seems to be able to build the toolchain a bit less flakily. In the meantime, macos CI will probably finish before that does, so I'll merge this PR before confirming if the macos toolchain build is fixed.

File: lib/Frontend/CompilerInvocation.cpp
Comment (rxwei) [reply lib/Frontend/CompilerInvocation.cpp]: I think you meant just a flag to `swiftc`? Yes, I think that would be better.

**Review by rxwei — APPROVED** [review]

**Comment by dbonner** [comment]
Hi,
My build fails with the following on macOS while trying to build swiftPM:
/Users/daniel/swift-source/swiftpm: error: manifest parse error(s):
swift (LLVM option parsing): for the --enable-experimental-cross-file-derivative-registration option: may only occur zero or one times!
Cheers,
Dan

**Comment by dan-zheng** [comment]
@dbonner: the issue should be fixed after https://github.com/apple/swift/pull/29156! I'm verifying it now.

**Pull request merged**
